subsys/shell/shell.c

```diff
@@ -1414,15 +1414,14 @@ void shell_uninit(const struct shell *sh, shell_uninit_cb_t cb)
 		/* signal kill message */
 		(void)k_poll_signal_raise(signal, 0);
 
-		return;
-	}
-
-	int err = instance_uninit(sh);
-
-	if (cb) {
-		cb(sh, err);
 	} else {
-		__ASSERT_NO_MSG(0);
+		int err = instance_uninit(sh);
+
+		if (cb) {
+			cb(sh, err);
+		} else {
+			__ASSERT_NO_MSG(0);
+		}
```

Comment:
Is this really necessary? Returning early from a function shouldn't require an additional level of nesting.

If the issue is that IAR doesn't allow local variables to be declared after control statements, then couldn't the declaration / initialization of the variable just be moved to the top of the function?

Alternatively, is there some flag that can be passed to IAR to allow for C99-style variable declaration? It would be a shame to have to regress to C89 to satisfy one compiler.

Reply:
It is only a warning as IAR defaults to C99. It only warns that the initialization might not occur due to an earlier return or goto, especially if it is dynamically initialized with a return value from a call with potential side effects.

We already have disabled quite a lot of warnings and errors to be able to compile zephyr. This one was just one that I thought wouldn't have to occur, but sure, it can be disabled.